internal/service/location/controller.go

```diff
@@ -85,7 +88,46 @@ func transformTollgates(tollgates []*tollgate.Tollgate) (result []*detector.Toll
 	return
 }
 
+func errorsToBadRequest(errors validation.Errors) *errdetails.BadRequest {
+	violations := make([]*errdetails.BadRequest_FieldViolation, 0)
+	for k, err := range errors {
+		violations = append(violations, &errdetails.BadRequest_FieldViolation{
+			Field:       k,
+			Description: err.Error(),
+		})
+	}
+	return &errdetails.BadRequest{
+		FieldViolations: violations,
+	}
+}
+
+type validator struct {
+	errors validation.Errors
+}
+
+func (v *validator) validate(name string, value interface{}, rules ...validation.Rule) {
+	err := validation.Validate(value, rules...)
+	if err != nil {
+		v.errors[name] = err
+	}
+}
+
 func (c *Controller) UpdateLocation(ctx context.Context, request *locationV1beta1.UpdateLocationRequest) (*locationV1beta1.UpdateLocationResponse, error) {
+	v := validator{errors: validation.Errors{}}
+	v.validate("worker_id", request.GetWorkerId(), validation.Required)
+	v.validate("timestamp", fmt.Sprintf("%v", request.GetTimestamp()), validation.Required, is.Int)
+	v.validate("lon", fmt.Sprintf("%v", request.GetLon()), validation.Required, is.Longitude)
+	v.validate("lat", fmt.Sprintf("%v", request.GetLat()), validation.Required, is.Latitude)
+
+	if len(v.errors) > 0 {
+		st, err := status.New(codes.InvalidArgument, "invalid request").
+			WithDetails(errorsToBadRequest(v.errors))
+		if err != nil {
+			panic(fmt.Errorf("enrich grpc status with details error: %w", err))
```

Comment:
Do we want to panic?

Reply:
it is recommended from pretty much all examples related to status.WithDetails wraps.

Reply:
OK. This panic is very improbable, right?

Reply:
it basically should-never-happen kind of error %-)